ophyd/indexed_device.py

```diff
@@ -0,0 +1,273 @@
+import collections
+import copy
+import itertools
+import string
+
+from .device import Device, Component, Kind, create_device_from_components
+from .utils import underscores_to_camel_case
+
+
+class _IndexedChildLevel(collections.abc.Mapping):
+    'A mapping helper to allow access of IndexedDevice[idx][idx2...]'
+    def __init__(self, device, mapping_dict):
+        self._device = device
+        self._d = mapping_dict
+
+    def __getitem__(self, item):
+        value = self._d[item]
+        if isinstance(value, str):
+            return getattr(self._device, value)
+        return _IndexedChildLevel(device=self._device, mapping_dict=value)
+
+    def __iter__(self):
+        yield from self._d
+
+    def __len__(self):
+        return len(self._d)
+
+    def __repr__(self):
+        options = set(str(s) for s in self)
+        return f'<{self.__class__.__name__} len={len(self)} options={options}>'
+
+
+class IndexedDevice(Device, collections.abc.Mapping):
+    '''
+    A Device subclass which allows for indexing of components (e.g., channels)
+    by using the getitem syntax (i.e., ``dev[idx]``)
+    '''
+
+    def __init__(self, prefix='', *, name, kind=None, read_attrs=None,
+                 configuration_attrs=None, parent=None, **kwargs):
+        self._root_index = _IndexedChildLevel(
+            self, mapping_dict=self._mapping_dict)
+        super().__init__(prefix=prefix, name=name, kind=kind,
+                         read_attrs=read_attrs,
+                         configuration_attrs=configuration_attrs,
+                         parent=parent, **kwargs)
+
+    def _instantiate_component(self, attr):
+        cpt = super()._instantiate_component(attr)
+        # Patch in the index on the component for reverse lookups:
+        cpt.index = self._attr_to_index[attr]
+        return cpt
+
+    def __getitem__(self, index):
+        return self._root_index[index]
+
+    def __iter__(self):
+        yield from self._root_index
+
+    def __len__(self):
+        return len(self._root_index)
+
+
+def _normalize_ranges(ranges):
+    'Normalize a range/str to be used with _apply_range_to_value'
+    if isinstance(ranges, (range, str)):
+        return [list(ranges)]
+
+    return [list(ri) for ri in ranges]
+
+
+def _apply_range_to_value(ranges, value):
+    'Get all values from a format string and the given range(s)'
+    return [
+        value.format(*item)
+        for item in itertools.product(*ranges)
+    ]
+
+
+def _mapping_dict_from_ranges(ranges, attrs, *, tuple_access=True):
+    'Create a dictionary to be used by _IndexedChildLevel'
+    mapping_dict = {}
+
+    for index, attr in zip(itertools.product(*ranges), attrs):
+        d = mapping_dict
+        for index_part in index[:-1]:
+            if index_part not in d:
+                d[index_part] = {}
```

Comment:
if allowing tuple access also want to stuff the index remaineder in at each level so `obj.inxd[0][('A', 'B')]` works?

Can we do this by making `__getitem__` smarter rather than pre-computing?

Reply:
I think pre-computing could potentially be less expensive as it happens only once on class definition.

Also, `__getitem__` dramatically increased in complexity in supporting the slice syntax, so unless there's much better reason, I'd prefer to keep it as-is.

Reply:
I too prefer fancy class setup to fancy `__getitem__`.